src/canard.c

```diff
@@ -1,6 +1,736 @@
+}
+
+/**
+ * Assigns a new node ID value to the current node.
+ */
+void canardSetLocalNodeID(CanardInstance* ins, uint8_t self_node_id)
+{
+  ins->node_id = self_node_id;
+}
+
+/**
+ * Returns node ID of the local node.
+ * Returns zero if the node ID has not been set.
+ */
+uint8_t canardGetLocalNodeID(const CanardInstance* ins)
+{
+  return ins->node_id;
+}
+
+/**
+ * Sends a broadcast transfer.
+ * If the node is in passive mode, only single frame transfers will be allowed.
+ */
+ int canardBroadcast(CanardInstance* ins,
+                      uint16_t data_type_id,
+                      uint8_t* inout_transfer_id,
+                      uint8_t priority,
+                      const void* payload,
+                      uint16_t payload_len)
+ {
+  if (payload == NULL)
+  {
+      return -1;
+  }
+  if (canardGetLocalNodeID(ins) == 0 && payload_len > 7)
+  {
+    return -1;
+  }
+  if (priority > 31)
+  {
+      return -1;
+  }
+
+  const uint32_t can_id = ((uint32_t)priority << 24) | ((uint32_t)data_type_id << 8) | (uint32_t)canardGetLocalNodeID(ins);
+  
+  canardEnqueueData(ins, can_id, inout_transfer_id, payload, payload_len);
+
+  tidIncrement(inout_transfer_id);
+
+  return 1;
+}
+
+/**
+ * Sends a request or a response transfer.
+ * Fails if the node is in passive mode.
+ */
+int canardRequestOrRespond(CanardInstance* ins,
+                            uint8_t destination_node_id,
+                            uint16_t data_type_id,
+                            uint8_t* inout_transfer_id,
+                            uint8_t priority,
+                            CanardRequestResponse kind,
+                            const void* payload,
+                            uint16_t payload_len)
+{
+  if (payload == NULL)
+  {
+      return -1;
+  }
+  if (canardGetLocalNodeID(ins) == 0)
+  {
+    return -1;
+  }
+  if (priority > 31)
+  {
+      return -1;
+  }
+
+  const uint32_t can_id = ((uint32_t)priority << 24) | ((uint32_t)data_type_id << 16) | 
+                            ((uint32_t)kind << 15) | ((uint32_t)destination_node_id << 8) | 
+                              (1 << 7) | (uint32_t)canardGetLocalNodeID(ins);
+
+  canardEnqueueData(ins, can_id, inout_transfer_id, payload, payload_len);
+
+  tidIncrement(inout_transfer_id);
+
+  return 1;
+}
+
+/**
+ * Returns a pointer to the top priority frame in the TX queue.
+ * Returns NULL if the TX queue is empty.
+ */
+const CanardCANFrame* canardPeekTxQueue(const CanardInstance* ins)
+{
+  if (ins->tx_queue == NULL)
+  {
+    return NULL;
+  }
+  return &ins->tx_queue->frame;
+}
+
+/**
+ * Removes the top priority frame from the TX queue.
+ */
+void canardPopTxQueue(CanardInstance* ins) 
+{
+  CanardTxQueueItem* item = ins->tx_queue;
+  ins->tx_queue = item->next;
+  canardFreeBlock(&ins->allocator, item);
+}
+
+/**
+ * Processes a received CAN frame with a timestamp.
+ */
+void canardHandleRxFrame(CanardInstance* ins, const CanardCANFrame* frame, uint64_t timestamp_usec)
```

Comment:
Let me add an explanation on how the missing feature - calling `canardShouldAcceptTransferPtr()` - should be implemented.

* If SOT is set, request the application via `canardShouldAcceptTransferPtr()` whether the transfer should be accepted. If it should, the application will also return the data type signature (see the updated API spec). The signature then should be stored in the RX state structure (better yet, not the whole signature, but rather its CRC-16 that will be used as a seed to the transfer CRC function. [Explanation](http://uavcan.org/Specification/4._CAN_bus_transport_layer/#multi-frame-transfer).)
* If SOT is not set (i.e. this frame continues another transfer), and if `canardRxStateTraversal()` could not find a transfer state (yes it should not be silently adding a new state if an existing one could not be found), ignore the frame.

Does this approach seem correct to everyone?